[#3] Add token specification (#4)
tqtezos/stablecoin: docs/specification.md

<!--
SPDX-FileCopyrightText: 2020 tqtezos
SPDX-License-Identifier: MIT
-->

**Overview**
------------

These specifications were assembled with the following references:

- [*CENTRE Fiat Token design*](https://github.com/centrehq/centre-tokens/blob/78d964a1a8d481ffd8152772d7a66e47df54b3db/doc/tokendesign.md) which is served as a reference specification to tezos stablecoin smart contract.

- Tezos Token Standard:
  [*FA2*](https://gitlab.com/tzip/tzip/-/blob/76d5f3791bfbfe3c9bf95ad5ec5fc6cbeeca2d0e/proposals/tzip-12/tzip-12.md) and
  [*ManagedLedger*](https://gitlab.com/tzip/tzip/blob/ae2f1e7ebb3454d811a2bea3cd0698b0e64ccea5/proposals/tzip-7/ManagedLedger.md)

- [*Michelson Contract Interfaces and Conventions*](https://gitlab.com/tzip/tzip/blob/ae2f1e7ebb3454d811a2bea3cd0698b0e64ccea5/proposals/tzip-4/tzip-4.md) TZIP which defines `view` and `void` type synonyms

**General Requirements**
------------------------

- The token(s) must be and FA2 compatible as to facilitate listing on
  exchanges and interaction with services which support FA2

**Ledger**
==========

Every address that is stored in ledger is associated with its

> current balance, token id and current minting allowance.

Reviewer: > token id

The contract should hold only one type of tokens. Storing `token id` for each address does not make sense, it is always 0. Some other places should be updated as well.

**masterMinter**

- Can add and remove minters.

- Can change minting allowance for any minter if that is present in
  its list of minters.

- Stores the list of current minters within itself, rather
  than exposing the map of master minter and its minters globally
  for the whole token.

**minter**

- Can create and destroy coins (that are allowed by their current
  minting allowance).

- Minting allowance is stored locally within the address
  allowing for multiple minters creating and destroying tokens.

**pauser**

- Can pause transferring, burning and minting operations.
  During the pause, these operations cannot be performed
  and fail with an error if the user decides to try them.

**blacklister (TBD)**

- Can blacklist a particular address preventing it from
  transferring, minting, burning and receiving tokens by
  removing them from whitelist.

**Token Functions**
===================

**Standard Token Functions**
----------------------------

Functions for the stablecoin token implementation which are common to the [*FA2 Tezos
Token Standard*](https://gitlab.com/tzip/tzip/-/blob/76d5f3791bfbfe3c9bf95ad5ec5fc6cbeeca2d0e/proposals/tzip-12/tzip-12.md). The

> token implementation in general follows an [*event loop*](https://en.wikipedia.org/wiki/Event_loop) pattern meaning that every parameter

Reviewer: AFAIU it only applies to getters that indeed take a callback contract. When you say that the token implementation in general follows an event loop, it makes one think that it applies to nearly everything that is going on in the contract.

Anyway, what's the purpose of this sentence? Passing a callback contract is a standard approach and basically the only way to get data from another contract in Tezos.

handles an additional callback call describing what needs to be done with
its result.

**transfer**

Types
```
tokenId = nat

transferParam
  ( address :from_
  , address :to_
  , tokenId :tokenId
  , nat     :value
  )

transfer = list transferParam
```

Parameter (in Michelson):
```
(list %transfer
  (pair
    (address %from_)
    (pair
      (address %to_)
      (pair
        (nat %tokenId)
        (nat %amount)
  )))
)
```

- Transfers given amounts of tokens between addresses.

- Each transfer must happen atomically, if one of them fails, then
  the whole transaction fail.

> - Transfers must follow permission policies that are described above.

Reviewer: "above"? Maybe below?

Types
```
tokenId = nat

operatorTokens =
  | AllTokens
  | SomeTokens (set tokenId)

operatorParam =
  ( address        :owner
  , address        :operator
  , operatorTokens :tokens
  )

isOperatorResponse =
  ( operatorParam :operator
  , bool          :isOperator
  )

isOperatorParam =
  ( operatorParam               :operator
  , contract isOperatorResponse :callback
  )

isOperator = isOperatorParam
```

Parameter (in Michelson):
```
(pair %isOperator
  (pair %operator
    (address %owner)
    (pair
      (address %operator)
      (or %tokens
        (unit %allTokens)
        (set %someTokens nat)
      )
  ))
  (contract %callback
    (pair
      (pair %operator
        (address %owner)
        (pair
          (address %operator)
          (or %tokens
            (unit %allTokens)
            (set %someTokens nat)
          )
      ))
      (bool %isOperator)
    )
  )
)
```

- Inspect if an address is an operator for the specified owner and
  token types.

- If the adddress is not an operator for at least one
  requested token type, then the result is `false`.

- It's possible to make this query for some specific tokens, or to
  all tokens.

**pause**

Types
```

> pause = (address :sender)

Reviewer: Why do we pass an address as an argument instead of just using the `SENDER` instruction?

```

Parameter (in Michelson)
```

> (pair %pause

Reviewer: Why is it a pair? That `unit` below is redundant.

  (address %sender)
  unit
)
```

- Pauses transferring, burning and minting operations so
  that they cannot be performed. All other operations remain
  unaffected.

> - Cannot be called multiple times.

Reviewer: I guess you mean that it can't be called multiple times without `unpause` call between `pause` calls. But can be called multiple times during contract's lifetime.

    )
  )
)
```

- Returns token pause status.

- *Note: It's not a part of ManagedLedger interface*.

**Custom token functions**
==========================

**Issuing and destroyng tokens**
--------------------------

**configureMinter**

Types
```
tokenId = nat

configureMinterParam =
  ( address :sender
  , address :minter
  , tokenId :tokenId
  )

configureMinter = configureMinterParam
```

Parameter (in Michelson)
```
(pair %configureMinter
  (address %sender)
  (pair
    (address %minter)
    (nat %tokenId)
))
```

- Adds `minter` to global minter list to allow him mint tokens of specific
  token type.

- Sender must be master minter.

- Minter cannot mint tokens untill `setMintingAllowance`
  is specified.

**removeMinter**

Types
```
tokenId = nat

removeMinter =
  ( address :sender
  , address :minter
  , tokenId :tokenId
  )
```

Parameter (in Michelson)
```
(pair %removeMinter
  (address %sender)
  (pair
    (address %minter)
    (nat %tokenId)
))
```

- Removes minter from the global minter list for some specific token type
  and sets its minting allowance to 0. Once minter is removed it will no
  longer be able to mint or burn tokens.

- Sender must be master minter.

> **setMintingAllowance**

Reviewer 2: See https://blog.smartdec.net/erc20-approve-issue-in-simple-words-a41aaf47bca6, maybe we should consider splitting it into two functions. Not a showstopper, ofc, but nice to _not_ have this thing mentioned in the audit report in future.

Author: IMSM we already had such solution earlier when I proposed to add `approveCAS`  method here 🤷‍♂

- Sender must be a blacklister.

**checkWhitelisted** (void address bool)

- Return a boolean value describing whether the
  given address is present in whitelist.

Role reassigning functions
==========================

**Owner**
---------

**transferOwnership**

Types
```
transferOwnership =
  ( address :owner
  , address :receiver
  )
```

Parameter (in Michelson):
```
(pair %transferOwnership
  (address %owner)
  (address %receiver)
)
```

- Set token owner to a new address.

- Sender must be current token owner.

- The current owner retains his priveleges up until
  `acceptOwnership` is called.

- Can be called multiple times, each call replaces pending
  owner with the new one. Note, that if proposed
  owner is the same as the current one, then the call
  is simply invalidated.

**getOwner**

Types
```
getOwner = (contract address :callback)
```

Parameter (in Michelson):
```
(pair %getOwner
  (contract %callback address)
)
```

- Returns current token owner address.

**acceptOwnership**

Types
```
acceptOwnership = unit
```

Parameter (in Michelson):
```
(unit %acceptOwnership)
```

- Accept ownership privileges.

- Sender must be a pending owner.

> **Admin**

Reviewer 2: Could you add "admin" to the list of roles and describe its capabilities?

Author: Actually we don't need it since due tot CENTRE specification it can only upgrade the contract which we agreed on not to for the time being

<!--
SPDX-FileCopyrightText: 2020 tqtezos
SPDX-License-Identifier: MIT
-->

**Overview**
------------

These specifications were assembled with the following references:

- [*CENTRE Fiat Token design*](https://github.com/centrehq/centre-tokens/blob/78d964a1a8d481ffd8152772d7a66e47df54b3db/doc/tokendesign.md) which is served as a reference specification to tezos stablecoin smart contract.

- Tezos Token Standard:
  [*FA2*](https://gitlab.com/tzip/tzip/-/blob/76d5f3791bfbfe3c9bf95ad5ec5fc6cbeeca2d0e/proposals/tzip-12/tzip-12.md) and
  [*ManagedLedger*](https://gitlab.com/tzip/tzip/blob/ae2f1e7ebb3454d811a2bea3cd0698b0e64ccea5/proposals/tzip-7/ManagedLedger.md)

- [*Michelson Contract Interfaces and Conventions*](https://gitlab.com/tzip/tzip/blob/ae2f1e7ebb3454d811a2bea3cd0698b0e64ccea5/proposals/tzip-4/tzip-4.md) TZIP which defines `view` and `void` type synonyms

**General Requirements**
------------------------

- The token(s) must be and FA2 compatible as to facilitate listing on
  exchanges and interaction with services which support FA2

**Ledger**
==========

Every address that is stored in ledger is associated with its

> current balance, current allowance and a list of minters.

Author: The point I was making is that any "additional" storage, associated to addresses in ledger is represented by `ledgerValue`. Actually I think so too that we need to remove `allowance` value since we do not implement any `approve` method that checks for its misuse (or should we?). 

Regarding a list of minters the point was that allow for multiple addresses to create and destroy contract tokens. For more you can refer to [CENTRE specification](https://github.com/centrehq/centre-tokens/blob/78d964a1a8d481ffd8152772d7a66e47df54b3db/doc/tokendesign.md#issuing-and-destroying-tokens) document.

Regarding explicit `senders` I think that we should leave them in specification so to clarify it (as in FA2 document). The `SENDER` approach is an implementation detail and I was going to stick to it anyway.

**Roles**
=========

The token supports the following list of user roles as is described in
CENTRE Fiat Token specification and the taxonomy of permission policies that are described in FA2 token specification.

<!-- Here're the roles that are present in current version of stablecoin token.  -->

Below we define some roles for stablecoin contract token.

**owner**

- Can re-assign any role of the token

**masterMinter**

- Can add and remove minters

- Can increase minting allowance for any minter

- Stores the list of current minters within itself, rather
  than exposing the map of master minter and its minters globally

>   for the whole token

Author: Yes, I meant that each `masterMinter` has a list of it's sub-minters that can receive tokens from their parent. All minters can depend only on one `masterMinter` preventing their cyclic appearances. Yes, it needs to be clarified that it can increase allowance for any minter within its list.

**minter**

- Can create and destroy coins (that are allowed by their current minting allowance)

- Minting allowance is stored locally within the address
  allowing for multiple minters creating and destroying tokens.

**pauser**

- Can pause transferring, burning and minting operations.
  During the pause, these operations cannot be performed

>   and remain unaffected

Author: Whops. Yes, they fail with an error if the contract is paused.

handles an additional callback call describing what needs to be done with
its result.

**transfer**

Types
```
tokenId = nat

transferParam
  ( address :from_
  , address :to_
  , tokenId :tokenId
  , nat     :value
  )

transfer = list transferParam
```

Parameter (in Michelson):
```
(list %transfer
  (pair
    (address %from_)
    (pair
      (address %to_)
      (pair
        (nat %tokenId)
        (nat %amount)
  )))
)
```

- Transfers given amounts of tokens between addresses.

- Each transfer must happen atomically, if one of them fails, then
  the whole transaction fail.

- Transfers must follow permission policies that are described above.

- The amount of a token transfer must not exceed the existing token
  owner's balance. If the transfer amount for the particular token
  type and token owner exceeds the existing balance, then the whole
  transfer operation fail.

- Core transfer behavior may be extended. If additional constraints on tokens
  transfer are required, FA2 token contract implementation may invoke
  additional permission policies. If the additional permission hook fails,

>   then the whole transfer fail.

Author: It's a part of [FA2 spec](https://gitlab.com/tzip/tzip/-/blob/76d5f3791bfbfe3c9bf95ad5ec5fc6cbeeca2d0e/proposals/tzip-12/tzip-12.md#transfer)

- Core transfer behavior must update token balances exactly as the
  operation parameters specify it. No changes to amount values or
  additional transfers are allowed.

> - Sender must be a minter.

Author: Actually it's not necessary)

operation on behalf of the `owner` that actually holds tokens.

**updateOperators**

Types
```
tokenId = nat

operatorTokens =
  | AllTokens
  | SomeTokens (set tokenId)

operatorParam =
  ( address        :owner
  , address        :operator
  , operatorTokens :tokens
  )

updateOperatorParam =
  | AddOperator    operatorParam
  | RemoveOperator operatorParam

updateOperators = list updateOperatorParam
```

Parameter (in Michelson)
```
(list %updateOperators
  (or
    (pair %addOperator
      (address %owner)
      (pair
        (address %operator)
        (or %tokens
          (unit %allTokens)
          (set %someTokens nat)
        )
    ))
    (pair %removeOperator
      (address %owner)
      (pair
        (address %operator)
        (or %tokens
          (unit %allTokens)
          (set %someTokens nat)
        )
    ))
  )
)
```

- Add or Remove token operators for the specified owners and token
  types.

- If two different commands in the list add and remove an operator
  for the same token type, then the last one must take effect.

- It's possible to update an operator for some specific tokens, or
  to all tokens (TODO: discuss).

> - Sender must be minter.

Author: At first I thought that it's be better to separate roles to minters so that we can assure that wallet has some balance before transferring. On a second thought it can be implemented easier just failing with an error if it does not have enough coins to transfer.

- Sender and operation receiver must be whitelisted.

- Contract must not be paused.

**approveOperatorUpdate (TODO)**

Types
```
tokenId

operatorTokens =
  | AllTokens
  | SomeTokens (set tokenId)

operatorParam =
  ( address        :owner
  , address        :operator
  , operatorTokens :tokens
  )

approveOperatorUpdate = operatorParam
```

Parameter (in Michelson):
```
(pair %approveOperatorUpdate
  (pair %operator
      (address %owner)
      (pair
        (address %operator)
        (or %tokens
          (unit %allTokens)
          (set %someTokens nat)
        )
    ))
  (contract %callback
    (pair %operator
      (address %owner)
      (pair
        (address %operator)
        (or %tokens
          (unit %allTokens)
          (set %someTokens nat)
        )
      )
    )
  )
)
```

- Enables `operator` to operate on tokens held by `owner`.

> - `owner` and `operator` must be whitelisted.

Author: Fair point

      )
  ))
  (contract %callback
    (pair
      (pair %operator
        (address %owner)
        (pair
          (address %operator)
          (or %tokens
            (unit %allTokens)
            (set %someTokens nat)
          )
      ))
      (bool %isOperator)
    )
  )
)
```

- Inspect if an address is an operator for the specified owner and    token types.

- If the adddress is not an operator for at least one
  requested token type, then the result is `false`.

- It's possible to make this query for some specific tokens, or to
  all tokens.

**Managed Ledger Functions (TBD)**
----------------------------

TODO: Maybe it's better to omit these mentions since all that left from
ManagedLedger is their semantics?

Functions for the TZUSDC token implementation found in Tezos [*ManagedLedger*](https://gitlab.com/tzip/tzip/blob/ae2f1e7ebb3454d811a2bea3cd0698b0e64ccea5/proposals/tzip-7/ManagedLedger.md)

**pause**

Types
```
pause = (address :sender)
```

Parameter (in Michelson)
```
(pair %pause
  (address %sender)
  unit
)
```

- Pauses transferring, burning, minting and allowance operations so
  that they cannot be performed. All other operations remain

>   unaffected.

Author: And how else this can be implemented then? If you're referring to a `map (address, nat)` so that represents allowance by default. If you mean `setAllowance`-like entrypoints then I think that it's better to be paused as well as other mentioned operations, but I'm not sure.

- Cannot be called multiple times.

- Sender must be a pauser.

**unpause**

Types
```
unpause = (address :sender)
```

Parameter (in Michelson)
```
(pair %unpause
  (address %sender)
  unit
)
```

- Unpauses the contract so that transferring, burning, minting and
  allowance operations can be performed by users with corresponding
  roles.

- Cannot be called multiple times.

- Sender must be a pauser.

**getPaused**

Types
```
tokenId = nat

getPausedResponse =
  ( tokenId :tokenId
  , nat     :paused
  )

getPausedParam =
  ( list tokenId                      :tokenIds
  , contract (list getPausedResponse) :callback
  )

getPaused = getPausedParam
```

Parameter (in Michelson)
```
(pair %getTotalPaused
  (list %tokenIds nat)
  (contract %callback
    (list
      (pair %getPausedResponse
        (nat %tokenId)
        (bool %paused)
      )
    )
  )
)
```

- Returns token pause status.

- *Note: It's not a part of ManagedLedger interface*.

**Custom token functions**
==========================

**Issuing and destroyng tokens**
--------------------------

**configureMinter**

Types
```
tokenId = nat

configureMinterParam =

>   ( address :sender

Reviewer 2: Why do you need `sender` field here? Are you going to check that the transaction sender is equal to the `sender` in the parameter? In this case, you can just use `SENDER` instruction to get the sender value directly.

Or do you want an _operator_ to be able to call this function? I wonder if people would wish to offload such a sensitive action to operators.

Author: Actually I don't think that operators should be allowed call anything but `transfer`. I pass sender address here to verify if the given address is `masterMinter` and has `minter` address inside its minter list (or rather set?)

Author: UPD: to check if sender address is master minter, I have already answered earlier about why I explicitly pass this in documentation even so when I'm going to use `SENDER` in actual implementation

  , address :minter
  , tokenId :tokenId
  )

configureMinter = configureMinterParam
```

Parameter (in Michelson)
```
(pair %configureMinter
  (address %sender)
  (pair
    (address %minter)
    (nat %tokenId)
))
```

- Adds `minter` to sender's minter list to allow him mint tokens of specific
  token type.

- Sender must master minter.

- Minter cannot mint tokens untill `setMintingAllowance`
  is specified.

> - Cannot add minter if present in someone else's minter list.

Reviewer 2: Implementation hint: it's better to have a global set (or, rather, `big_map address ()`) of all minters so that you can check this quickly without iterating over master minters' minters lists.

**removeMinter**

Types
```
tokenId = nat

removeMinter =

>   ( address :sender

Reviewer 2: Same question regarding the sender as in `configureMinter`

  , address :minter
  , tokenId :tokenId
  )
```

Parameter (in Michelson)
```
(pair %removeMinter
  (address %sender)
  (pair
    (address %minter)
    (nat %tokenId)
))
```

- Removes minter from the sender's minter list for some specific token type
  and sets its minting allowance to 0. Once minter is removed it will no
  longer be able to mint or burn tokens.

- Sender must be master minter.

- Minter address must present in sender's minter list.

**setMintingAllowance**

Types
```
tokenId = nat

setMintingAllowanceParam =

>   ( address :sender

Reviewer 2: Same question again (see `configureMinter`)

          (address %owner)
          (nat %tokenId)
        )
        (nat %allowance)
      )
    )
  )
)
```

- Returns current minting allowance for address with a specific token type.

**getTotalMinted**

Types
```
tokenId = nat

getTotalMintedResponse =
  ( tokenId :tokenId
  , nat     :totalMinted
  )

getTotalMintedParam =
  ( list tokenId                           :tokenIds
  , contract (list getTotalMintedResponse) :callback
  )

getTotalMinted = getTotalMintedParam
```

Parameter (in Michelson)
```
(pair %getTotalMinted
  (list %tokenIds nat)
  (contract %callback
    (list
      (pair %getTotalMintedResponse
        (nat %tokenId)
        (nat %totalMinted)
      )
    )
  )
)
```
- Returns current amount of minted coins for an address with a specific token
  type.

-- TODO: clarify under which conditions and how should
-- token reset this allowance

**setMintingAllowanceResetInterval** timestamp

Types
```
setMintingAllowanceResetInterval =
  ( tokenId   :tokenId
  , timestamp :interval
  )

```

Parameter (in Michelson)
```
(pair %setMintingAllowanceResetInterval
  (address %sender)
  (timestamp %interval)
)
```

- Set the default interval of minting allowance reset.

> - Set to 0 to make disable minting allowance reset by stablecoin token.

Reviewer 2: I propose to use `option` here but use your judgment

Author: I think it would be better to stick to FA2 spec as much as possible

**mint (TBD)**

Types
```
tokenId = nat

mintParam
  ( address :sender

>   , address :minter

Reviewer 2: You can use `SENDER` to get the transaction sender and check whether it is a `minter`.
I don't get why you use `minter` here at all because AFAIU minter is the sender in all cases.

But I don't get why you do not include the token recipient in the transaction param. Who will receive the tokens in this case?

Author: Yep, it should be renamed to `:masterMinter` and `:recipient`.

  , tokenId :tokenId
  , nat     :value
  )

mint = list mintParam
```

Parameter (in Michelson):
```
(list %mint
  (pair
    (address %sender)
    (pair
      (address %minter)
      (pair
        (nat %tokenId)
        (nat %value)
  )))
)
```

- Produces the given amounts of tokens to the wallets associated with the
  given addresses.

> - Each minting must happen atomically, si if one of them fails, then the whole

Reviewer 2: ```suggestion
- Each minting must happen atomically, so if one of them fails, then the whole
```

  operation must fail.

- The operation must follow permission policies described above.

- Receiving addresses must be whitelisted.

> - Sender must be master minter.

Reviewer 2: ```suggestion
- Sender must be minter.
```

- The total amount of minted coins must not exceed the current

>   minting allowance specified for each recepient individually.

Reviewer 2: ```suggestion
  minting allowance specified for each minter individually.
```

- Minting allowance will decrease by the amount of tokens minted
  and increase the balance of receiver and total minted value.

- Contract must not be paused.

**burn (TBD)**

Types
```
tokenId = nat

burnParam
  ( address :sender

>   , address :minter

Reviewer 2: Same question about params here

  , tokenId :tokenId
  , nat     :value
  )

burn = list burnParam
```

Parameter (in Michelson):
```
(list %burn
  (pair
    (address %sender)
    (pair
      (address %minter)
      (pair
        (nat %tokenId)
        (nat %value)
  )))
)
```

- Decreases balances for specified wallets and their token types and the total   supply of tokens by the given amount.

> - Each minting must happen atomically, si if one of them fails, then the whole

Reviewer 2: ```suggestion
- Each burn must happen atomically, so if one of them fails, then the whole
```

  operation must fail.

- The operation must follow permission policies described above.

- Sender must be minter and the receiver must have a sufficient amount of

>   funds to be destroyed and be in a sender's minter list.

Reviewer 2: AFAIU minter can burn his tokens _only._ Do I get something wrong? The CENTRE token spec says:
>  The minter specifies the amount of tokens to burn, and the minter must have a balance greater than or equal to the amount. Burning tokens is restricted to minter addresses to avoid accidental burning of tokens by end users.

- A minter with 0 minting allowance is allowed to burn tokens.

- Contract must not be paused.

**Whitelisting (TBD)**
----------------

> Functions for the TZUSDC token implementation which are outside the FA1.2 Tezos Token Standard.

Reviewer 2: ```suggestion
Whitelisting functions for the TZUSDC token implementation which are outside the FA2 Tezos Token Standard.
```

Author: Haven't touched this section at all

**addToWhitelist** address

- Add an address to whitelist, enabling it to transfer and
  approve tokens as well as modify allowances.

- Sender must be a blacklister.

**addToWhitelistBatch** (list address)

- Adds each address to whitelist, enabling it to transfer and
  approve tokens as well as modify allowances.

- Sender must be a blacklister.

**removeFromWhitelist** address

- Remove address from whitelist.

- Sender must be a blacklister.

**removeFromWhitelistBatch** (list address)

- Remove provided addresses from whitelist.

- If some of specified addresses are not in whitelist, then
  it fails returning list of those addresses.

- Sender must be a blacklister.

**checkWhitelisted** (void address bool)

- Return a boolean value describing whether the
  given address is present in whitelist.

Role reassigning functions
==========================

**Owner**
---------

**transferOnership**

Types
```
transferOwnership =

>   ( address :owner

Reviewer 2: You can use `SENDER` value and compare it against the current `owner`

  , address :receiver
  )
```

Parameter (in Michelson):
```
(pair %transferOwnership
  (address %owner)
  (address %receiver)
)
```

- Set token owner to a new address.

- Sender must be current token owner.

- The current owner retains his priveleges up until
  `acceptOwnership` is called.

- Can be called multiple times, each call replaces pending
  owner with the new one. Note, that if proposed
  owner is the same as the current one, then the call
  is simply invalidated.

**getOwner**

Types
```
getOwner = (contract address :callback)
```

Parameter (in Michelson):
```
(pair %getOwner
  (contract %callback address)
)
```

- Returns current token owner address.

**acceptOwnership**

Types
```
acceptOwnership = unit
```

Parameter (in Michelson):
```
(unit %acceptOwnership)
```

- Accept ownership privileges.

- Sender must be a pending owner.

**Admin**
---------

**changeTokenAdmin** address

Types
```
changeTokenAdmin =

>   ( address :owner

Reviewer 2: Same here

  , address :receiver
  )
```

Parameter (in Michelson):
```
(pair %changeTokenAdmin
  (address %owner)
  (address %receiver)
)
```

- Set token administrator to a new address.

- Sender must be current token admin.

- The current administrator retains his priveleges up until
  `acceptAdminRole` is called.

- Can be called multiple times, each call replaces pending
  administrator with the new one. Note, that if proposed
  administrator is the same as the current one, then the call
  is simply invalidated.

**getAdmin**

Types
```
getAdmin = (contract address :callback)
```

Parameter (in Michelson):
```
(pair %getAdmin
  (contract %callback address)
)
```

- Returns current token administrator address.

**acceptAdminRole**

Types
```
acceptAdminRole = unit
```

Parameter (in Michelson):
```
(unit %acceptAdminRole)
```

- Accept administrator privileges.

- Sender must be a pending administrator.

**Master Minter**
---------------

**changeMasterMinter**

Types
```
changeMasterMinter =

>   ( address :owner

Reviewer 2: Same here

  , address :receiver
  )
```

Parameter (in Michelson):
```
(pair %changeMasterMinter
  (address %owner)
  (address %receiver)
)
```

> - Set master minter to a new address.

Reviewer 2: But earlier in the spec you assumed that there would be several master minters. Have you abandoned this idea?

Author: Nope, I'll clarify this

**owner**

- Can re-assign any role of the token.

**masterMinter**

- Can add and remove minters.

- Can increase minting allowance for any minter if that is present in 
  its list of minters.

- Stores the list of current minters within itself, rather
  than exposing the map of master minter and its minters globally

>   for the whole token.

Author: Yes
